packages/runtime/container-runtime/src/dataStoreContext.ts

```diff
@@ -662,24 +662,33 @@ export class RemotedFluidDataStoreContext extends FluidDataStoreContext {
                     throw new Error(`Invalid snapshot format version ${snapshotFormatVersion}`);
                 }
                 this.pkg = pkgFromSnapshot;
-                isRootStore = isRootDataStore;
             }
 
-            /**
-             * If there is no isRootDataStore in the attributes blob, set it to true. This will ensure that
-             * data stores in older documents are not garbage collected incorrectly. This may lead to additional
-             * roots in the document but they won't break.
-             */
             this.details = {
                 // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                 pkg: this.pkg!,
-                isRootDataStore: isRootStore ?? true,
                 snapshot: tree ?? undefined,
             };
         }
 
         return this.details;
     }
+
+    protected async getInitialGCDetails(): Promise<IGCDetails> {
+        if (this.initialGCDetails === undefined) {
```

Comment:
getInitialSnapshotDetails() is guaranteed to be called once, because it is only called by realizeCore, and it's wrapper with realize() and deferral.
isRoot(), however, can be called 20 times in a row, and each of them will result in being stuck waiting for readAndParse() below. I think what we need is for this.initialGCDetails not to be results, but promise to result, and this function not having awaits, but rather using continuations to immediately set this.initialGCDetails to pending promise.

Reply:
Good point!

Reply:
Changed this to LazyPromise which seems to fit pretty well with what we are trying to do here.
Also, since isRootDataStore is moved back to getInitialSnapshotDetails(), its no longer guaranteed that it will only be called by realize. So, I changed that to a lazy promise too.